### PR TITLE
fix(pwa): enable offline compose shell for /to/<user> routes and background sync

### PR DESCRIPTION
**Pull request by nandhakumar-murugan**
Hi @kgashok sir,

Thank you for testing this and highlighting this crucial workflow!

### Root Cause Analysis
1. **Missing Offline Compose Precache:** `/to/<user>` routes were not precached in the Service Worker, and offline navigation previously fell back to the homepage (`/`) rather than an offline compose shell.
2. **External Editor Dependencies:** The ToastUI Markdown Editor JS/CSS and helper scripts were missing from `STATIC_ASSETS`, preventing the form from rendering when offline.

### The Fix (Submitted in PR #540)
1. **Offline Compose Shell (`/compose-offline`):** Added a lightweight, generic fallback route in `saythanks/core.py` and precached it along with all ToastUI editor bundles and jQuery dependencies in `saythanks/static/sw.js` (`saythanks-v3`).
2. **Dynamic Client-Side Route Parsing:** In `submit_note.htm.j2`, when served from the offline cache, the client-side script parses `window.location.pathname` to extract the recipient slug (e.g. `lifebalance`) and topic, dynamically updating the title, form action (`./lifebalance/submit`), and editor greeting.
3. **SQLAlchemy 2.0 Compatibility:** Updated `saythanks/storage.py` with thread-safe parameter mapping across worker threads.
4. **Automated Verification:** Added `tests/test_pwa_offline.py` (10/10 tests pass).

### Verification Evidence

**1. Offline Mode (`127.0.0.1:5000/to/lifebalance`):**  
Navigated while completely offline &mdash; the compose form rendered immediately from cache without any browser dinosaur/owl error, and clicking "Send Note" queued it in the offline outbox (`📬 Outbox: 1 pending sync`).

Evidence 1 - Offline Compose

**2. Reconnection Sync:**  
As soon as connectivity was restored, `offline-outbox.js` automatically flushed the queue, sending the POST request (`submit` 302 &rarr; `thanks` 200) without requiring any user re-submission.

Evidence 2 - Reconnect Sync

Follow-up PR is open and ready for your review: **#540**. Thank you! 🚀

### CODE REVIEW

**Comment by nandhakumar-murugan** [comment]
### ✅ Addressed GitHub Copilot Review Feedback (Commit `f1a11fa`)

All issues flagged in Copilot's review have been resolved, verified, and pushed in commit `f1a11fa`:

1. **Explicit Commit for RETURNING Statements (`saythanks/storage.py`)**:
   - `CompatConnection.execute()` now unconditionally calls `current_conn.commit()` so statements using `INSERT / UPDATE ... RETURNING` (such as `Note.store()`) persist transactions reliably before the connection context closes.

2. **Safe JavaScript JSON Escaping (`saythanks/templates/submit_note.htm.j2`)**:
   - Replaced string interpolation with Jinja's safe `tojson` filter (`{{ user|tojson }}` and `{{ topic|tojson }}`) to prevent script breakage or injection vulnerabilities.

3. **Unconditional Core Dependencies (`saythanks/core.py`)**:
   - Restored unconditional imports for hard runtime dependencies including `Markup`, `Common`, `get_full_name`, `Sentry`, `QRcode`, and `Cleaner`.

4. **MailerSend & Versioning (`saythanks/myemail.py`, `saythanks/version.py`)**:
   - Restored standard unconditional imports for `mailersend` matching upstream.
   - Updated `test_email_templates.py` mocking to supply `__version__` so test execution remains clean and isolated.

5. **Precache Resilience & Runtime CDN Caching (`saythanks/static/sw.js`)**:
   - Removed external CDN URLs (`uicdn.toast.com`, `ajax.googleapis.com`) from install-time `STATIC_ASSETS` to avoid precache failures on network issues or first-time offline loads.
   - Preserved Strategy A runtime cache-first fallback for these CDN hosts on subsequent fetches.

6. **Unconditional Tablib Import (`saythanks/storage.py`)**:
   - Restored standard unconditional `import tablib` to prevent `AttributeError` during inbox data export.

7. **Conditional Auth0 Client Initialization (`saythanks/storage.py`)**:
   - Initialized `auth0` only when both `auth0_domain` and `auth0_token` (`AUTH0_JWT_V2_TOKEN`) are non-empty: `auth0 = Auth0(...) if (auth0_domain and auth0_token) else None`.

8. **Clean Local SQLite Configuration (`saythanks/storage.py`)**:
   - Removed hardcoded default SQLite fallback in production paths; only ensures tables locally when SQLite is explicitly active.

9. **Removed Startup Seeding Side Effects (`saythanks/storage.py`)**:
   - Removed all startup/import-time `INSERT OR IGNORE` seed rows to eliminate side effects across environments.

10. **Narrowed Transaction Exceptions (`saythanks/storage.py`)**:
    - Replaced generic `Exception` aliases with distinct fallback classes (`InFailedSqlTransaction`, `UniqueViolation`) when `psycopg2` is not installed.

### Automated Test Suite
All 10 tests pass cleanly:
```text
tests/test_email_templates.py ...                                        [ 30%]
tests/test_pwa_offline.py ...                                            [ 60%]
tests/test_submit_note_template.py ....                                  [100%]
======================== 10 passed, 1 warning in 1.08s ========================
```
Ready for re-review! 🚀

**Comment by kgashok** [comment]
@nandhakumar-murugan I am not accepting this PR for now since the previous PR has been rejected. Combine code from your previous PR along with this one. And if possible, provide multiple PRs. For example, provide a separate PR only for the test cases. 

Remember, you cannot blindly submitted all the code that AI generates. Please don't submit code that you don't understand, yet. 

Ideally, you must first provide a proposal before you even submit a PR. That would be better for me to review your approach and then accept your PR. For a good example, talk to @Sudarshan-prog - he provided [this](https://github.com/BlitzKraft/saythanks.io/blob/master/docs/PR_418_Architecture_Documentation.md) to resolve #418

**Comment by kgashok** [comment]
@nandhakumar-murugan Just a question: For the offline store of the saythanks note, are you using sqlite? And then using sqlite to update the postgres when the app goes online?

**Comment by nandhakumar-murugan** [comment]
Hi @kgashok  sir,

No, SQLite is not used for offline storage or syncing. Offline notes are stored in the browser using IndexedDB (offline-outbox.js) and replayed directly to the Flask backend (saving to PostgreSQL) once back online. SQLite was only added for local development testing.

Understood on the proposal and split PR approach. I will work on this and update you, sir!

**Comment by kgashok** [comment]
> Understood on the proposal and split PR approach. I will work on this and update you, sir!

@nandhakumar-murugan and @harrish243 , Review the commits I have made here in https://github.com/BlitzKraft/saythanks.io/pull/543 and https://github.com/BlitzKraft/saythanks.io/pull/545 - and get familiar with the code that got added. And then test out the current version (Sept 9 release). Your task (s) now is submit issues based on what you see is right or wrong with the current offline implementation.
